### ISSUE TITLE
flatpak-system-helper grows in memory when chain installing apps (potentially leaking)

### ISSUE THREAD

**Issue by ahayzen** [flatpak/flatpak#1700]
## Linux distribution and version

Ubuntu 18.04

## Flatpak version

Flatpak 0.11.7
OSTree 2018.4

## Description of the problem

When chain installing multiple apps I happened to be watching the highest memory of processes (both "memory" and resident memory in gnome-system-monitor). I noticed that the process `flatpak-system-helper` increased in memory with each app that it installed (this potentially could affect updates as well).

It seems like with each application it doesn't clean up and the memory increases in steps, eg starting low ~50-100, then 100-200, 200-300 etc with each app.

## Steps to reproduce

1) In a terminal run `$ flatpak install flathub org.kde.kdenlive org.kde.krita org.kde.okular` eg pick at least 3 apps that you don't have installed (also I was missing at least one runtime)
2) Watch in your system monitor the memory of flatpak-system-helper increase as each app in turn is installed.

## What I expected to happen

At step 2 for the memory to remain consistent and not increase as more apps are installed.

## Valgrind

I ran the steps under `valgrind` by starting `# valgrind /usr/lib/flatpak/flatpak-system-helper --replace --no-idle-exit` as root before step 1.

flatpak install - https://pastebin.ubuntu.com/p/75G4Nqpnbk/
valgrind  flatpak-system-helper - https://pastebin.ubuntu.com/p/HWFqG2XPwc/

The valgrind summary at exit is
```
==20103== HEAP SUMMARY:
==20103==     in use at exit: 6,519,331 bytes in 43,738 blocks
==20103==   total heap usage: 21,881,673 allocs, 21,837,935 frees, 1,045,276,412 bytes allocated
==20103== 
==20103== LEAK SUMMARY:
==20103==    definitely lost: 2,047,160 bytes in 6,941 blocks
==20103==    indirectly lost: 621,269 bytes in 9,571 blocks
==20103==      possibly lost: 2,960 bytes in 26 blocks
==20103==    still reachable: 3,823,854 bytes in 27,010 blocks
==20103==                       of which reachable via heuristic:
==20103==                         length64           : 3,064 bytes in 64 blocks
==20103==                         newarray           : 1,952 bytes in 42 blocks
==20103==         suppressed: 0 bytes in 0 blocks
```

Which doesn't suggest much is definitely lost (~2MiB) however I wonder if the memory is not being "lost" but just incorrectly held between the steps of installing each app ?

**Comment by ahayzen** [comment]
@matthiasclasen this is the output with `--leak-check=full --verbose`  https://pastebin.ubuntu.com/p/2sPbMppx4K/  I installed some of the missing debug symbols, eg for libsoup, ostree etc let me know if you need more.

Note that this is on ubuntu with flatpak 0.11.7 and ostree 2018.4 (if you need to match the line numbers).

**Comment by matthiasclasen** [comment]
Looking at ostree git logs, the big leak you're seeing might be fixed by:

commit 9721be34e161ff5b2d699c52e1add633e8c33347
Author: Matthew Leeds <matthew.leeds@endlessm.com>
Date:   Wed Mar 28 23:11:50 2018 -0700

    libotutil/checksum-utils: Fix memory management
    
    Ostree uses the OtChecksum data structure as a wrapper around GChecksum
    (depending on what libraries are available at compile time). According
    to the docs for g_checksum_get_digest(), a GChecksum value can no longer
    be updated after that function is called. Ostree enforces this by
    setting "initialized" to FALSE after getting the digest, but this leads
    to ot_checksum_clear() avoiding freeing any memory, leading to leaks. So
    this commit adds a "closed" value that gets set when getting a digest
    and checked when updating the value, so the initialized value can be
    used only for memory management.
    
    Closes: #1521
    Approved by: jlebon

**Comment by matthiasclasen** [comment]
Closing this for now until new evidence appears.
